recipes_source/generic_join.rst

```diff
@@ -0,0 +1,434 @@
+- A dive into how the ``Join`` context manager works.
+- An example showing how to make a toy class compatible with the context
+  manager.
+
+Requirements
+------------
+
+- PyTorch 1.10+
+- `Getting Started with Distributed Data Parallel`_
+- `Shard Optimizer States with ZeroRedundancyOptimizer`_
+
+What is ``Join``?
+-----------------
+In `Getting Started with Distributed Data Parallel - Basic Use Case`_, you saw
+the general skeleton for using ``DistributedDataParallel`` to perform data
+parallel training. This implicitly schedules all-reduces in each backward pass
+to synchronize gradients across ranks. Such `collective communications
+<https://pytorch.org/docs/stable/distributed.html>`__ require participation
+from all ranks in the process group, so if a rank has fewer inputs, then the
+other ranks will hang or error (depending on the backend). More generally, this
+problem persists for any class that performs per-iteration synchronous
+collective communications.
+
+``Join`` is a context manager to be used around your per-rank training loop to
+facilitate training with uneven inputs. The context manager allows the ranks
+that exhaust their inputs early (i.e. *join* early) to shadow the collective
+communications performed by those that have not yet joined. The ways in which
+the communications are shadowed are specified by hooks.
+
+Using ``Join`` with ``DistributedDataParallel``
+-----------------------------------------------
+PyTorch's ``DistributedDataParallel`` works out-of-the-box with the ``Join``
+context manager. Here is an example usage:
+
+::
+
+    import os
+    import torch
+    import torch.distributed as dist
+    import torch.multiprocessing as mp
+    from torch.distributed.algorithms.join import Join
+    from torch.nn.parallel import DistributedDataParallel as DDP
+
+    BACKEND = "nccl"
+    WORLD_SIZE = 2
+    NUM_INPUTS = 5
+
+    def worker(rank):
+        os.environ['MASTER_ADDR'] = 'localhost'
+        os.environ['MASTER_PORT'] = '29500'
+        dist.init_process_group(BACKEND, rank=rank, world_size=WORLD_SIZE)
+
+        model = DDP(torch.nn.Linear(1, 1).to(rank), device_ids=[rank])
+        # Rank 1 gets one more input than rank 0
+        inputs = [torch.tensor([1]).float() for _ in range(NUM_INPUTS + rank)]
+
+        num_inputs = 0
+        with Join([model]):
+            for input in inputs:
+                num_inputs += 1
+                loss = model(input).sum()
+                loss.backward()
+
+        print(f"Rank {rank} has exhausted all {num_inputs} of its inputs!")
+
+    def main():
+        mp.spawn(worker, nprocs=WORLD_SIZE, join=True)
+
+    if __name__ == "__main__":
+        main()
+
+This produces the following output (where the ``print()`` s from rank 0 and
+rank 1 may be arbitrarily ordered):
+
+::
+
+  Rank 0 has exhausted all 5 of its inputs!
+  Rank 1 has exhausted all 6 of its inputs!
+
+.. note::
+    ``DistributedDataParallel`` provided its own ``join()`` context manager
```

Comment:
Let's convert `DistributedDataParallel` and DDP's `join()` into hyperlinks as well.